### PR TITLE
Make capsys more like stdio streams in python3.  Resolves #1407.

### LINKED ISSUE #1407: capsys + sys.stdout.buffer => AttributeError
Applicable versions where sys.stdout.buffer should exist and be a bytestream:
- python3.x

``` python
import sys


def test(capsys):
    sys.stdout.buffer.write(b'hi\n')
    assert capsys.readouterr()[0] == 'hi\n'
```

```
$ py.test test.py
============================= test session starts ==============================
platform linux -- Python 3.4.3, pytest-2.8.7, py-1.4.31, pluggy-0.3.1
rootdir: /home/asottile/workspace/cheetah_lint, inifile: 
collected 1 items 

test.py F

=================================== FAILURES ===================================
_____________________________________ test _____________________________________

capsys = <_pytest.capture.CaptureFixture object at 0x7f15ca2cd320>

    def test(capsys):
>       sys.stdout.buffer.write(b'hi\n')
E       AttributeError: '_io.StringIO' object has no attribute 'buffer'

test.py:5: AttributeError
=========================== 1 failed in 0.01 seconds ===========================
```

### PR DESCRIPTION
**Pull request by asottile**
Thanks for submitting a PR, your contribution is really appreciated!

Here's a quick checklist that should be present in PRs:

- [x] Target: for bug or doc fixes, target `master`; for new features, target `features`;

Unless your change is trivial documentation fix (e.g.,  a typo or reword of a small section) please:

- [x] Make sure to include one or more tests for your change;
- [x] Add yourself to `AUTHORS`;
- [x] Add a new entry to `CHANGELOG.rst`
  * Choose any open position to avoid merge conflicts with other PRs.
  * Add a link to the issue you are fixing (if any) using RST syntax.
  * The pytest team likes to have people to acknowledged in the `CHANGELOG`, so please add a thank note to yourself ("Thanks @user for the PR") and a link to your GitHub profile. It may sound weird thanking yourself, but otherwise a maintainer would have to do it manually before or after merging instead of just using GitHub's merge button. This makes it easier on the maintainers to merge PRs.

### CODE REVIEW

**Comment by asottile** [comment]
This seems to break python3.3, which is a bit perplexing!  Lemme see if I can't figure out why (though python3.3 is getting pretty old (end of lifed in 2014)).

**Comment by asottile** [comment]
Neat: I fixed py33 by making `pytester` also use `_pytest.capture` instead of `py.io`.  It seems there was a mysterious refcounting bug that only affects python3.3 that `py.io` was tickling.

**Comment by The-Compiler** [comment]
Not sure if this should go to `features` instead.

**Comment by asottile** [comment]
Is it not a bug that capsys incorrectly mocks stdout / stderr (why I decided to PR master)?

**Comment by The-Compiler** [comment]
It probably is - however, it seems to me like there's quite some potential of breakage in some corner-cases here (like the one you already saw above :wink:), and we're trying to avoid regressions in patch releases. History shows that usually, something weird ends up breaking without the pytest testsuite catching it, as pytest has a lot of users :wink:

However, I don't have any strong feelings about this - let's see what others think.

**Comment by flub** [comment]
Looks fine to me.

**Comment by nicoddemus** [comment]
@asottile could you please take a look at the Windows tests? 

I agree with @The-Compiler that we should merge this into `features` instead, for the reasons he mentioned.

**Comment by asottile** [comment]
Fixed windows, it seems a little silly to be afraid to release bug fixes but it's not my project so I'll do as you please.  I'm currently blocked on this fix, when is your next release?

**Comment by nicoddemus** [comment]
Hmm can you provide an example which demonstrates the problem without the fix?

@RonnyPfannschmidt do you think this should go into `features` or it could go into `master`?

**Comment by asottile** [comment]
There's a couple in my original issue: https://github.com/pytest-dev/pytest/issues/1407

File: testing/test_capture.py [hunk 1]
```diff
@@ -642,6 +642,13 @@ def test_unicode_and_str_mixture(self):
             f.close()
             assert isinstance(s, unicode)
 
+    def test_write_bytes_to_buffer(self):
+        f = capture.CaptureIO()
+        # In python3, stdout / stderr are text io wrappers (exposing a buffer
+        # property of the underlying bytestream).
+        getattr(f, 'buffer', f).write(b'foo\n')
```
Comment (The-Compiler) [inline testing/test_capture.py]: This seems like it'd pass even if the object doesn't have a `buffer` attribute (i.e. without any code changes).

Comment (asottile) [inline testing/test_capture.py]:
```python
>>> import io
>>> f = io.StringIO()
>>> getattr(f, 'buffer', f).write(b'foo\n')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: string argument expected, got 'bytes'
```

Nope! Thankfully `io.StringIO` *requires* text (the prior implementation)

**Comment by asottile** [comment]
I found a workaround with `capfd`.  This can go into features instead.  I'll close this in favor of my other PR

**Pull request closed without merging**